File: codes/Aiamjay/Week4-Day3/62. Unique Paths.py
# encoding = utf-8


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[1] * m for _ in range(n)]
        for row in range(1, n):
            for col in range(1, m):
                dp[row][col] = dp[row - 1][col] + dp[row][col - 1]
        return dp[-1][-1]

    def formula(self, m, n) -> int:
        total_steps = m + n - 2
        down_steps = n - 1
        # wjcnote the ans is C(total_steps, down_steps)  C=combination组合
        ans = 1
        for i in range(1, down_steps):
            ans *= (total_steps - down_steps + i) / i
        return int(ans)

    def test_solution(self):
        # case 1
        print(self.uniquePaths(7, 3))

        # corner case
        print(self.uniquePaths(1, 1))

        # corner case
        print(self.uniquePaths(1, 10))
        pass


if __name__ == '__main__':
    s = Solution()
    s.test_solution()

File: codes/Shuofei/week 4/62_Unique Paths.java
class Solution {
    public int uniquePaths(int m, int n) {
        // C(m + n, n) == C(m + n, m)        
        int t = Math.max(m, n);
        double res = 1;
        for (int i = m + n - 2; i >= t; i--)
            res = res * i / (m + n - 1 - i);
        return (int)res;
    }
}

class Solution {
    public int uniquePaths(int m, int n) {
        int[][] f = new int[m + 1][n + 1];
        f[1][1] = 1;
        for (int i = 1; i <= m; i++)
            for (int j = 1; j <= n; j++)
                if (!((i == 1) && (j == 1)))
                    f[i][j] = f[i-1][j] + f[i][j-1];
        return f[m][n];
    }
}